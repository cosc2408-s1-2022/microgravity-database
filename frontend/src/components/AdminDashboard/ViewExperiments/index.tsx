import {
  DeleteRounded,
  DoneRounded,
  EditRounded,
  NewReleasesRounded,
  RestartAltRounded,
  VerifiedSharp,
  WarningRounded,
} from '@mui/icons-material';
import {
  Link,
  Box,
  Button,
  Grid,
  Pagination,
  Paper,
  Typography,
  Tooltip,
  useTheme,
  useMediaQuery,
} from '@mui/material';
import { AxiosResponse } from 'axios';
import moment from 'moment';
import { ChangeEvent, useEffect, useState } from 'react';
import { useMutation, useQuery, useQueryClient } from 'react-query';
import { useNavigate } from 'react-router-dom';
import api from '../../../util/api';
import { Experiment, ResultsResponse } from '../../../util/types';
import CenteredNoneFound from '../../CenteredNoneFound';
import MessageSnackbar from '../../MessageSnackbar';

type ViewExperimentsProps = {
  page?: number;
  size?: number;
  searchString?: string;
  onPageChange: (_e: ChangeEvent<unknown>, page: number) => void;
};

export default function ViewExperiments({ page, size, searchString, onPageChange }: ViewExperimentsProps) {
  const navigate = useNavigate();
  const queryClient = useQueryClient();

  const [experiments, setExperiments] = useState<ResultsResponse<Experiment>>();
  const {
    data: experimentsData,
    isSuccess: isExperimentsSuccess,
    isLoading: hasPendingRequests,
    isError: isExperimentsError,
    refetch: refetchExperiments,
  } = useQuery<AxiosResponse<ResultsResponse<Experiment>>>(
    ['getExperiments', page, size, searchString],
    ({ queryKey, signal }) => {
      const [, page, size, searchString] = queryKey;
      const params = new URLSearchParams();
      searchString && params.append('string', searchString as string);
      page && params.append('page', page as string);
      size && params.append('size', size as string);

      const paramsEncoded = encodeURI(params.toString());
      const url = searchString
        ? `/search?${paramsEncoded}`
        : `/experiments/paginated${paramsEncoded !== '' ? `?${paramsEncoded}` : ''}`;

      return api.get(url, { signal });
    },
    {
      enabled: false,
    },
  );
  useEffect(() => {
    if (hasPendingRequests) queryClient.cancelQueries('getExperiments');
    refetchExperiments();
  }, [refetchExperiments, page, size, searchString, hasPendingRequests, queryClient]);
  useEffect(() => {
    if (isExperimentsSuccess && experimentsData) {
      setExperiments(experimentsData.data);
    }
  }, [isExperimentsSuccess, experimentsData]);

  const {
    isLoading: isToggleDeleteLoading,
    isSuccess: isToggleDeleteSuccess,
    mutate: toggleExperimentDelete,
  } = useMutation('toggleExperimentDelete', (id: number) => api.post(`/experiments/${id}/toggleDelete`));
  useEffect(() => {
    if (isToggleDeleteSuccess) {
      refetchExperiments();
    }
  }, [isToggleDeleteSuccess, refetchExperiments]);

  const {
    isLoading: isApproveLoading,
    isSuccess: isApproveSuccess,
    mutate: approveExperiment,
  } = useMutation('approveExperiment', (id: number) => api.post(`/experiments/${id}/approve`));
  useEffect(() => {
    if (isApproveSuccess) {
      refetchExperiments();
    }
  }, [isApproveSuccess, refetchExperiments]);

  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up('md'));

  return (
    <Grid container spacing={2} mb={3}>
      {experiments?.totalElements === 0 ? (
        <Grid item xs={12}>
          <CenteredNoneFound />
        </Grid>
      ) : (
        experiments?.results.map((e) => (
          <Grid item key={e.id} xs={12}>
            <Paper
              sx={{
                display: 'flex',
                alignItems: 'center',
                flexWrap: 'nowrap',
                p: 1,
                boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)',
              }}
            >
              <Box display='flex' alignItems='center' flexGrow={1}>
                {e.approved && !e.deleted ? (
                  <Tooltip title='This experiment has been approved by an admin.'>
                    <VerifiedSharp fontSize='medium' color='success' />
                  </Tooltip>
                ) : e.deleted ? (
                  <Tooltip title='This experiment has been deleted and will not be publicly visible.'>
                    <WarningRounded fontSize='medium' color='warning' />
                  </Tooltip>
                ) : (
                  <Tooltip title='This experiment requires approval.'>
                    <NewReleasesRounded fontSize='medium' color='error' />
                  </Tooltip>
                )}
                <Link ml={1} href={`/experiment/${e.id}`}>
                  <Typography variant='body2' fontWeight='bold' color={e.deleted ? 'text.secondary' : 'text.primary'}>
                    {e.title} {e.deleted && ' (DELETED)'}
                  </Typography>
                  <Typography variant='body2' color='text.secondary' flexGrow={1} pr={2}>
                    {e.mission.name} &bull; Added {moment(e.createdAt).fromNow()}
                  </Typography>
                </Link>
              </Box>
              <Box display='flex' ml={2} justifyContent='flex-end' alignItems='center' sx={{ width: '20rem' }}>
                {!e.approved && !e.deleted && (
                  <Button
                    onClick={() => {
                      approveExperiment(e.id);
                    }}
                    variant='contained'
                    color='primary'
                    disabled={isApproveLoading}
                    sx={matches ? { mr: 1, width: '7rem' } : { mr: 1 }}
                  >
                    {matches && (
                      <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                        Approve
                      </Typography>
                    )}
                    <DoneRounded fontSize='small' />
                  </Button>
                )}
                <Button
                  onClick={() => {
                    navigate('/admin/experiments/edit', {
                      state: e,
                    });
                  }}
                  variant='contained'
                  color='primary'
                  sx={matches ? { mr: 1, width: '5rem' } : { mr: 1 }}
                >
                  {matches && (
                    <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                      Edit
                    </Typography>
                  )}
                  <EditRounded fontSize='small' />
                </Button>
                <Button
                  onClick={() => {
                    toggleExperimentDelete(e.id);
                  }}
                  disabled={isToggleDeleteLoading}
                  variant='contained'
                  color='primary'
                  sx={matches ? { mr: 1, width: '7rem' } : { mr: 1 }}
                >
                  {matches && (
                    <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                      {e.deleted ? 'Restore' : 'Delete'}
                    </Typography>
                  )}
                  {e.deleted ? <RestartAltRounded fontSize='small' /> : <DeleteRounded fontSize='small' />}
                </Button>
              </Box>
            </Paper>
          </Grid>
        ))
      )}
      <Grid item xs={12} display='flex' justifyContent='center'>
        <Pagination
          count={experiments?.totalPages}
          variant='outlined'
          shape='rounded'
          color='secondary'
          page={page || 1}
          siblingCount={0}
          onChange={onPageChange}
        />
      </Grid>
      <MessageSnackbar
        open={isExperimentsError}
        message='Could not load experiments. Please try again.'
        severity='error'
      />
      <MessageSnackbar open={isApproveSuccess} message='Experiment approved.' severity='success' />
    </Grid>
  );
}
