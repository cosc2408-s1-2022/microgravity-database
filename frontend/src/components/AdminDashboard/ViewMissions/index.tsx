import {
  VerifiedSharp,
  WarningRounded,
  NewReleasesRounded,
  DoneRounded,
  EditRounded,
  DeleteRounded,
  RestartAltRounded,
} from '@mui/icons-material';
import {
  Box,
  Button,
  Grid,
  Link,
  Pagination,
  Paper,
  Tooltip,
  Typography,
  useMediaQuery,
  useTheme,
} from '@mui/material';
import { AxiosResponse } from 'axios';
import moment from 'moment';
import { ChangeEvent, useState, useEffect } from 'react';
import { useMutation, useQuery, useQueryClient } from 'react-query';
import { useNavigate } from 'react-router-dom';
import api from '../../../util/api';
import { ResultsResponse, Mission } from '../../../util/types';
import CenteredNoneFound from '../../CenteredNoneFound';
import MessageSnackbar from '../../MessageSnackbar';
import lodash from 'lodash';

type ViewMissionProps = {
  page?: number;
  size?: number;
  searchString?: string;
  onPageChange: (_e: ChangeEvent<unknown>, page: number) => void;
};

export default function ViewMissions({ page, size, searchString, onPageChange }: ViewMissionProps) {
  const navigate = useNavigate();
  const queryClient = useQueryClient();

  const [missions, setMissions] = useState<ResultsResponse<Mission>>();
  const {
    data: missionsData,
    isSuccess: isMissionsSuccess,
    isLoading: hasPendingRequests,
    isError: isMissionsError,
    refetch: refetchMissions,
  } = useQuery<AxiosResponse<ResultsResponse<Mission>>>(
    ['getMissions', page, size, searchString],
    ({ queryKey, signal }) => {
      const [, page, size, searchString] = queryKey;
      const params = new URLSearchParams();
      searchString && params.append('string', searchString as string);
      page && params.append('page', page as string);
      size && params.append('size', size as string);

      const paramsEncoded = encodeURI(params.toString());
      const url = searchString
        ? `/search/missions?${paramsEncoded}`
        : `/missions/paginated${paramsEncoded !== '' ? `?${paramsEncoded}` : ''}`;

      return api.get(url, { signal });
    },
    {
      enabled: false,
    },
  );
  useEffect(() => {
    if (hasPendingRequests) queryClient.cancelQueries('getMissions');
    refetchMissions();
  }, [refetchMissions, page, size, searchString, hasPendingRequests, queryClient]);
  useEffect(() => {
    if (isMissionsSuccess && missionsData) {
      setMissions(missionsData.data);
    }
  }, [isMissionsSuccess, missionsData]);

  const {
    isLoading: isToggleDeleteLoading,
    isSuccess: isToggleDeleteSuccess,
    mutate: toggleMissionDelete,
  } = useMutation('toggleMissionDelete', (id: number) => api.post(`/missions/${id}/toggleDelete`));
  useEffect(() => {
    if (isToggleDeleteSuccess) {
      refetchMissions();
    }
  }, [isToggleDeleteSuccess, refetchMissions]);

  const {
    isLoading: isApproveLoading,
    isSuccess: isApproveSuccess,
    mutate: approveMission,
  } = useMutation('approveMission', (id: number) => api.post(`/missions/${id}/approve`));
  useEffect(() => {
    if (isApproveSuccess) {
      refetchMissions();
    }
  }, [isApproveSuccess, refetchMissions]);

  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up('md'));

  return (
    <Grid container spacing={2} mb={3}>
      {missions?.totalElements === 0 ? (
        <Grid item xs={12}>
          <CenteredNoneFound />
        </Grid>
      ) : (
        missions?.results.map((m) => (
          <Grid item key={m.id} xs={12}>
            <Paper
              sx={{
                display: 'flex',
                alignItems: 'center',
                flexWrap: 'nowrap',
                p: 1,
                boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)',
              }}
            >
              <Box display='flex' alignItems='center' flexGrow={1}>
                {m.approved && !m.deleted ? (
                  <Tooltip title='This mission has been approved by an admin.'>
                    <VerifiedSharp fontSize='medium' color='success' />
                  </Tooltip>
                ) : m.deleted ? (
                  <Tooltip title='This mission has been deleted and will not be publicly visible.'>
                    <WarningRounded fontSize='medium' color='warning' />
                  </Tooltip>
                ) : (
                  <Tooltip title='This mission requires approval.'>
                    <NewReleasesRounded fontSize='medium' color='error' />
                  </Tooltip>
                )}
                <Link ml={1} href={`/mission/${m.id}`}>
                  <Typography variant='body2' fontWeight='bold' color={m.deleted ? 'text.secondary' : 'text.primary'}>
                    {m.name} {m.deleted && ' (DELETED)'}
                  </Typography>
                  <Typography variant='body2' color='text.secondary' flexGrow={1} pr={2}>
                    {moment(m.launchDate).year()} &bull; {lodash(m.platform.name).startCase()} &bull; Added{' '}
                    {moment(m.createdAt).fromNow()}
                  </Typography>
                </Link>
              </Box>
              <Box display='flex' ml={2} justifyContent='flex-end' alignItems='center' sx={{ width: '20rem' }}>
                {!m.approved && !m.deleted && (
                  <Button
                    onClick={() => {
                      approveMission(m.id);
                    }}
                    variant='contained'
                    color='primary'
                    disabled={isApproveLoading}
                    sx={matches ? { mr: 1, width: '7rem' } : { mr: 1 }}
                  >
                    {matches && (
                      <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                        Approve
                      </Typography>
                    )}
                    <DoneRounded fontSize='small' />
                  </Button>
                )}
                <Button
                  onClick={() => {
                    navigate('/admin/missions/edit', {
                      state: m,
                    });
                  }}
                  variant='contained'
                  color='primary'
                  sx={matches ? { mr: 1, width: '5rem' } : { mr: 1 }}
                >
                  {matches && (
                    <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                      Edit
                    </Typography>
                  )}
                  <EditRounded fontSize='small' />
                </Button>
                <Button
                  onClick={() => {
                    toggleMissionDelete(m.id);
                  }}
                  disabled={isToggleDeleteLoading}
                  variant='contained'
                  color='primary'
                  sx={matches ? { mr: 1, width: '7rem' } : { mr: 1 }}
                >
                  {matches && (
                    <Typography variant='body1' color='primary' textTransform='none' mr={1}>
                      {m.deleted ? 'Restore' : 'Delete'}
                    </Typography>
                  )}
                  {m.deleted ? <RestartAltRounded fontSize='small' /> : <DeleteRounded fontSize='small' />}
                </Button>
              </Box>
            </Paper>
          </Grid>
        ))
      )}
      <Grid item xs={12} display='flex' justifyContent='center'>
        <Pagination
          count={missions?.totalPages}
          variant='outlined'
          shape='rounded'
          color='secondary'
          page={page || 1}
          siblingCount={0}
          onChange={onPageChange}
        />
      </Grid>
      <MessageSnackbar open={isMissionsError} message='Could not load missions. Please try again.' severity='error' />
      <MessageSnackbar open={isApproveSuccess} message='Mission approved.' severity='success' />
    </Grid>
  );
}
