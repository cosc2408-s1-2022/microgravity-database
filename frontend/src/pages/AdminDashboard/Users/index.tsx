import {
  Container,
  FormControl,
  Grid,
  OutlinedInput,
  InputLabel,
  InputAdornment,
  IconButton,
  Select,
  MenuItem,
  Divider,
  SelectChangeEvent,
} from '@mui/material';
import { ChangeEvent, useEffect, useState } from 'react';
import { Navigate, useNavigate } from 'react-router-dom';
import CenteredCircularProgress from '../../../components/CenteredCircularProgress';
import { UserRole } from '../../../util/types';
import { useLoggedInUser } from '../../../util/hooks';
import { ClearRounded, SearchRounded } from '@mui/icons-material';
import UserSearch from '../../../components/UserSearch';

export default function Users() {
  const navigate = useNavigate();

  const { user: loggedInUser, isLoading: isLoggedInUserLoading, isError: isLoggedInUserError } = useLoggedInUser();
  useEffect(() => {
    if (isLoggedInUserError) {
      navigate('/login');
    }
  }, [isLoggedInUserError, navigate]);

  const [size, setSize] = useState<number>();
  const [searchString, setSearchString] = useState<string>();
  const handleSearchInputChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setSearchString(e.target.value);
  };
  const handleSearchInputClick = () => {
    if (searchString) {
      setSearchString(undefined);
    }
  };
  const handleSizeChange = (e: SelectChangeEvent<string | number>) => {
    setSize(e.target.value as number);
  };

  return isLoggedInUserLoading || !loggedInUser ? (
    <CenteredCircularProgress />
  ) : loggedInUser.role === UserRole.ROLE_ADMIN ? (
    <Container maxWidth='sm' sx={{ mt: 4 }}>
      <Grid container spacing={2}>
        <Grid item xs={12} display='flex' justifyContent='space-between' alignItems='center'>
          <FormControl size='small' variant='outlined'>
            <InputLabel htmlFor='search-input' color='secondary'>
              Search
            </InputLabel>
            <OutlinedInput
              id='search-input'
              type='text'
              value={searchString || ''}
              onChange={handleSearchInputChange}
              label='Search'
              color='secondary'
              endAdornment={
                <InputAdornment position='end'>
                  <IconButton onClick={handleSearchInputClick} edge='end'>
                    {searchString ? <ClearRounded /> : <SearchRounded />}
                  </IconButton>
                </InputAdornment>
              }
            />
          </FormControl>
          <FormControl size='small' variant='outlined'>
            <InputLabel id='select-size' color='secondary'>
              Result Size
            </InputLabel>
            <Select
              labelId='select-size'
              value={size || ''}
              label='Result Size'
              color='secondary'
              sx={{ minWidth: '8rem' }}
              onChange={handleSizeChange}
            >
              <MenuItem value={8}>Default</MenuItem>
              <MenuItem value={16}>16</MenuItem>
              <MenuItem value={32}>32</MenuItem>
              <MenuItem value={96}>96</MenuItem>
            </Select>
          </FormControl>
        </Grid>
        <Grid item xs={12}>
          <Divider />
        </Grid>
        <Grid item xs={12}>
          <UserSearch size={size} searchString={searchString} loggedInUser={loggedInUser} />
        </Grid>
      </Grid>
    </Container>
  ) : (
    <Navigate to='/login' />
  );
}
