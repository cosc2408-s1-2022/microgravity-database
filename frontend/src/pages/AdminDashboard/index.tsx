import { Box, Container, Tab, Tabs } from '@mui/material';
import { useEffect, useState } from 'react';
import { Outlet, useNavigate } from 'react-router-dom';

export default function AdminDashboard() {
  const navigate = useNavigate();
  const [value, setValue] = useState(0);
  useEffect(() => {
    switch (value) {
      case 0:
        navigate('users');
        break;
      case 1:
        navigate('experiments');
        break;
      default:
        break;
    }
  }, [value, navigate]);

  return (
    <Container maxWidth='lg' sx={{ mt: 2, display: 'flex', flexDirection: 'column', justifyContent: 'space-evenly' }}>
      <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
        <Tabs value={value} onChange={(_e, v) => setValue(v)} indicatorColor='secondary' textColor='secondary'>
          <Tab label='Users' sx={{ fontSize: 'large', color: 'text.primary', textTransform: 'none' }} />
          <Tab label='Experiments' sx={{ fontSize: 'large', color: 'text.primary', textTransform: 'none' }} />
        </Tabs>
      </Box>
      <Outlet />
    </Container>
  );
}
