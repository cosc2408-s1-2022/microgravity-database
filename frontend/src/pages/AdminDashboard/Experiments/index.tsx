import {
  Container,
  Divider,
  FormControl,
  Grid,
  IconButton,
  InputAdornment,
  InputLabel,
  MenuItem,
  OutlinedInput,
  Select,
  SelectChangeEvent,
} from '@mui/material';
import { ChangeEvent, useState } from 'react';
import { ClearRounded, SearchRounded } from '@mui/icons-material';
import AuthWrapper from '../../../components/AuthWrapper';
import ViewExperiments from '../../../components/AdminDashboard/ViewExperiments';

export default function Experiments() {
  const [size, setSize] = useState<number>();
  const [searchString, setSearchString] = useState<string>();
  const handleSearchInputChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setSearchString(e.target.value);
  };
  const handleSearchInputClick = () => {
    if (searchString) {
      setSearchString(undefined);
    }
  };
  const handleSizeChange = (e: SelectChangeEvent<string | number>) => {
    setSize(e.target.value as number);
  };

  return (
    <AuthWrapper>
      <Container maxWidth='lg' sx={{ mt: 3 }}>
        <Grid container spacing={2}>
          <Grid item xs={12} display='flex' justifyContent='space-between' alignItems='center'>
            <FormControl size='small' variant='outlined'>
              <InputLabel htmlFor='search-input' color='secondary'>
                Search
              </InputLabel>
              <OutlinedInput
                id='search-input'
                type='text'
                value={searchString || ''}
                onChange={handleSearchInputChange}
                label='Search'
                color='secondary'
                endAdornment={
                  <InputAdornment position='end'>
                    <IconButton onClick={handleSearchInputClick} edge='end'>
                      {searchString ? <ClearRounded /> : <SearchRounded />}
                    </IconButton>
                  </InputAdornment>
                }
              />
            </FormControl>
            <FormControl size='small' variant='outlined'>
              <InputLabel id='select-size' color='secondary'>
                Result Size
              </InputLabel>
              <Select
                labelId='select-size'
                value={size || ''}
                label='Result Size'
                color='secondary'
                sx={{ minWidth: '8rem' }}
                onChange={handleSizeChange}
              >
                <MenuItem value={8}>Default</MenuItem>
                <MenuItem value={16}>16</MenuItem>
                <MenuItem value={32}>32</MenuItem>
                <MenuItem value={96}>96</MenuItem>
              </Select>
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <Divider />
          </Grid>
          <Grid item xs={12}>
            <ViewExperiments size={size} searchString={searchString} />
          </Grid>
        </Grid>
      </Container>
    </AuthWrapper>
  );
}
