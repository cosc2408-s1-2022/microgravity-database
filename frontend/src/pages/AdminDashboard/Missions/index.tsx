import { ClearRounded, SearchRounded } from '@mui/icons-material';
import {
  Container,
  Divider,
  FormControl,
  Grid,
  IconButton,
  InputAdornment,
  InputLabel,
  MenuItem,
  OutlinedInput,
  Select,
  SelectChangeEvent,
} from '@mui/material';
import { useState, ChangeEvent, useEffect } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import ViewMissions from '../../../components/AdminDashboard/ViewMissions';
import AuthWrapper from '../../../components/AuthWrapper';
import { UserRole } from '../../../util/types';

export default function Missions() {
  const navigate = useNavigate();
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);

  const pageParam = searchParams.get('page');
  const sizeParam = searchParams.get('size');
  const searchStringParam = searchParams.get('string');

  const [page, setPage] = useState<number | undefined>(pageParam ? Number(pageParam) : undefined);
  const [size, setSize] = useState<number | undefined>(sizeParam ? Number(sizeParam) : undefined);
  const [searchString, setSearchString] = useState<string | undefined>(searchStringParam ?? undefined);
  const handlePageChange = (_e: ChangeEvent<unknown>, page: number) => {
    setPage(page);
  };
  const handleSizeChange = (e: SelectChangeEvent<string | number>) => {
    setSize(e.target.value as number);
    setPage(undefined);
  };
  const handleSearchInputChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setSearchString(e.target.value);
    setPage(undefined);
  };
  const handleSearchInputClick = () => {
    if (searchString) {
      setSearchString(undefined);
      setPage(undefined);
    }
  };

  useEffect(() => {
    const params = new URLSearchParams();
    page && params.append('page', page.toString());
    size && params.append('size', size.toString());
    searchString && params.append('string', searchString);

    const paramsEncoded = encodeURI(params.toString());
    paramsEncoded && navigate(`?${paramsEncoded}`);
  }, [page, size, searchString, navigate]);

  return (
    <AuthWrapper role={UserRole.ROLE_ADMIN}>
      <Container maxWidth='lg' sx={{ mt: 2 }}>
        <Grid container spacing={2}>
          <Grid item xs={12} display='flex' justifyContent='space-between' alignItems='center'>
            <FormControl size='small' variant='outlined'>
              <InputLabel htmlFor='search-input' color='secondary'>
                Search
              </InputLabel>
              <OutlinedInput
                id='search-input'
                type='text'
                value={searchString || ''}
                onChange={handleSearchInputChange}
                label='Search'
                color='secondary'
                endAdornment={
                  <InputAdornment position='end'>
                    <IconButton onClick={handleSearchInputClick} edge='end'>
                      {searchString ? <ClearRounded /> : <SearchRounded />}
                    </IconButton>
                  </InputAdornment>
                }
              />
            </FormControl>
            <FormControl size='small' variant='outlined'>
              <InputLabel id='select-size' color='secondary'>
                Result Size
              </InputLabel>
              <Select
                labelId='select-size'
                value={size || ''}
                label='Result Size'
                color='secondary'
                sx={{ minWidth: '8rem' }}
                onChange={handleSizeChange}
              >
                <MenuItem value={8}>Default</MenuItem>
                <MenuItem value={16}>16</MenuItem>
                <MenuItem value={32}>32</MenuItem>
                <MenuItem value={96}>96</MenuItem>
              </Select>
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <Divider />
          </Grid>
          <Grid item xs={12}>
            <ViewMissions page={page} size={size} searchString={searchString} onPageChange={handlePageChange} />
          </Grid>
        </Grid>
      </Container>
    </AuthWrapper>
  );
}
