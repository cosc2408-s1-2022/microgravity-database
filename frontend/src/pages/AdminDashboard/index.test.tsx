import React from 'react';
import { cleanup } from '@testing-library/react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from 'react-query';
import AdminDashboard from './index';

afterEach(cleanup);
describe('Rendering page Admin Dashboard page', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    const queryClient = new QueryClient();
    ReactDOM.render(
      <BrowserRouter>
        <QueryClientProvider client={queryClient}>
          <AdminDashboard />
        </QueryClientProvider>
      </BrowserRouter>,
      div,
    );
    expect(div).toMatchSnapshot();
  });
});
