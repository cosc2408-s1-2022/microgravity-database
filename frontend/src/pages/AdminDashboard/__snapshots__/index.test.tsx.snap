// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering page Admin Dashboard page renders without crashing 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-10s5e0p-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1gsv261"
    >
      <div
        class="MuiTabs-root css-1ujnqem-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-kbubpg-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Users
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-kbubpg-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Experiments
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-kbubpg-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Missions
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-kbubpg-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="-1"
              type="button"
            >
              People
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
